parser_hevc.c

```diff
@@ -0,0 +1,387 @@
+      pkt->data     = m_PesBuffer;
+      pkt->dts      = m_DTS;
+      pkt->pts      = m_PTS;
+      pkt->duration = duration;
+      pkt->streamChange = streamChange;
+
+    }
+    m_StartCode = 0xffffffff;
+    m_PesParserPtr = 0;
+    m_FoundFrame = false;
+    m_FrameValid = true;
+  }
+}
+
+void cParserHEVC::Reset()
+{
+  cParser::Reset();
+  m_StartCode = 0xffffffff;
+  m_NeedSPS = true;
+  m_NeedPPS = true;
+  memset(&m_streamData, 0, sizeof(m_streamData));
+  m_rbspBufferSize = 0;
+  if (m_rbspBuffer)
+  {
+    free(m_rbspBuffer);
+    m_rbspBuffer = NULL;
+  }
+}
+
+int cParserHEVC::Parse_HEVC(int buf_ptr, bool &complete)
+{
+  unsigned int len = m_PesBufferPtr - buf_ptr;
+  uint8_t *buf = m_PesBuffer + buf_ptr;
+  unsigned int p=0, NumBytesInNalUnit=0;
+
+  // find length of nal (including nal_unit_header)
+  uint32_t startcode = 0xffffffff;
+  while ((p+3) < len)
+  {
+    if ((startcode & 0x00ffffff) == 0x00000001)
+    {
+      NumBytesInNalUnit = p-3;  
+      break;
+    }
+    startcode = startcode << 8 | buf[p++];
+  }
+  if (!NumBytesInNalUnit)
+    return -1;
+
+  // create buffer for rbsp
+  if (NumBytesInNalUnit > m_rbspBufferSize)
+  {
+    if (NumBytesInNalUnit >= 500000)
+    {
+      ERRORLOG("cParserHEVC::Parse_HEVC - max buffer size reached, pid: %d", m_pID);
+      Reset();
+      return 0;
+    }
+    m_rbspBufferSize = NumBytesInNalUnit;
+    uint8_t *new_buffer = (uint8_t*)realloc(m_rbspBuffer, m_rbspBufferSize);
+    if (new_buffer == NULL)
+    {
+      ERRORLOG("cParserHEVC::Parse_HEVC - realloc failed");
+      Reset();
+      return 0;
+    }
+    m_rbspBuffer = new_buffer;
+    INFOLOG("HEVC rbspBuffer resized to %i", NumBytesInNalUnit);
+  }
+
+  // nal_unit_header
+  uint16_t header = (buf[0] << 8) | buf[1];
+  if (header & 0x8000) // ignore forbidden_bit == 1
+    return 0;
+  hevc_private::HDR_NAL hdr;
+  hdr.nal_unit_type = (header & 0x7e00) >> 9;
+  hdr.nuh_layer_id = (header & 0x1f8) >> 3;
+  hdr.nuh_temporal_id = (header & 7) - 1;
+
+  // extract SODB from RBSP: 7.3.1.1 General NAL unit syntax
+  unsigned int NumBytesInRbsp = 0, i=2;
+  while (i < NumBytesInNalUnit-2)
+  {
+    if ((buf[i] == 0) && (buf[i+1] == 0) && (buf[i+2] == 3))
+    {
+      m_rbspBuffer[NumBytesInRbsp++] = buf[i++];
+      m_rbspBuffer[NumBytesInRbsp++] = buf[i++];
+      i++; //skip emulation_prevention_three_byte
+    }
+    else
+      m_rbspBuffer[NumBytesInRbsp++] = buf[i++];
+  } 
+  m_rbspBuffer[NumBytesInRbsp++] = buf[i++];
+  m_rbspBuffer[NumBytesInRbsp++] = buf[i++];
+
+
+  switch (hdr.nal_unit_type)
+  {
+  case NAL_TRAIL_N ... NAL_RASL_R:
+  case NAL_BLA_W_LP ... NAL_CRA_NUT:
+  {
+    if (m_NeedSPS || m_NeedPPS)
+    {
+      m_FoundFrame = true;
+      return 0;
+    }
+    hevc_private::VCL_NAL vcl;
+    memset(&vcl, 0, sizeof(hevc_private::VCL_NAL));
+    if (!Parse_SLH(m_rbspBuffer, NumBytesInRbsp, hdr, vcl))
+      return 0;
+ 
+    // check for the beginning of a new access unit
```

Comment:
this is all about 7.4.2.4.3 Order of access units and association to CVSs
Did you match this with the spec?